src/v/rpc/parse_utils.h

```diff
@@ -115,6 +115,49 @@ concept is_rpc_serde_exempt = requires {
     typename T::rpc_serde_exempt;
 };
 
+/*
+ * Encode a client request for the given transport version.
+ *
+ * Unless the message type T is explicitly exempt from adl<> support, type T
+ * must be supported by both adl<> and serde encoding frameworks. When the type
+ * is not exempt from adl<> support, serde is used when the version >= v2.
+ *
+ * The returned version indicates what level of encoding is used. This is always
+ * equal to the input version, except for serde-only messags which return v2.
+ * Callers are expected to further validate the runtime implications of this.
+ */
+template<typename T>
+ss::future<transport_version>
+encode_for_version(iobuf& out, T msg, transport_version version) {
+    static_assert(
+      (!is_rpc_adl_exempt<T> && !is_rpc_serde_exempt<T>)
+      || !(is_rpc_adl_exempt<T> && is_rpc_serde_exempt<T>));
```

Comment:
`(!a && !b) || !(a && b)` simplifies as:
`!(a && b)` or
`!a || !b`

Which isn't what you wrote in the comment, but I guess it's implied that some serialization is required!
```suggestion
    static_assert(!is_rpc_adl_exempt<T> || !is_rpc_serde_exempt<T>)
```
Which I read as: "If T is not explicitly adl<> exempt, T must support serde<>". But then I think that the requirement is a little weak, since we also have `is_serde_compatible_v`

Reply:
Maybe it's this:
```c++
requires(!is_rpc_adl_exempt<T> || serde::is_serde_compatible_v<T>)
```

Reply:
everything has to be serde _and_ adl supported, unless its exempted, but it can only be exempt form one. i don't think there is much reason to check if they are compatible--only exempt, since if they aren't exempt and arent' compaible, it won't compile.

Reply:
I think the concept check generally has nicer error message, and removes the function from overload resolution, but I'm like a 2/10 on this.